packages/lib/server/service/hashedLinks.service.ts

```diff
@@ -0,0 +1,219 @@
+    const currentLinks = normalizedLinks.map((l) => l.link);
+
+    // Delete links that are no longer in the list
+    const linksToDelete = connectedMultiplePrivateLinks.filter((link) => !currentLinks.includes(link));
+    await this.hashedLinksRepository.deleteLinks(eventTypeId, linksToDelete);
+
+    // Create or update links
+    for (const linkData of normalizedLinks) {
+      const exists = connectedMultiplePrivateLinks.includes(linkData.link);
+      if (!exists) {
+        await this.hashedLinksRepository.createLink(eventTypeId, linkData);
+      } else {
+        await this.hashedLinksRepository.updateLink(eventTypeId, linkData);
+      }
+    }
+  }
+
+  /**
+   * Validates a link and increments usage count if valid
+   * Handles both time-based and usage-based expiration with timezone awareness
+   * @param linkId - The hashed link ID to validate
+   * @returns The validated link data
+   * @throws Error with ErrorCode.PrivateLinkExpired if link is expired or invalid
+   */
+  async validateAndIncrementUsage(linkId: string) {
+    if (!linkId || typeof linkId !== "string") {
+      throw new Error("Invalid link ID");
+    }
+
+    const hashedLink = await this.hashedLinksRepository.findLinkWithValidationData(linkId);
+
+    if (!hashedLink) {
+      throw new Error(ErrorCode.PrivateLinkExpired);
+    }
+
+    // Handle time-based expiration with timezone awareness
+    if (hashedLink.expiresAt) {
+      const hostTimezone = this.extractHostTimezone(hashedLink.eventType);
+
+      if (hostTimezone) {
+        // Use dayjs for timezone-aware comparison
+        const now = dayjs().tz(hostTimezone);
+        const expiration = dayjs(hashedLink.expiresAt).tz(hostTimezone);
+
+        if (expiration.isBefore(now)) {
+          throw new Error(ErrorCode.PrivateLinkExpired);
+        }
+      } else {
+        // Fallback to UTC comparison if no timezone available
+        const now = dayjs();
+        const expiration = dayjs(hashedLink.expiresAt);
+
+        if (expiration.isBefore(now)) {
+          throw new Error(ErrorCode.PrivateLinkExpired);
+        }
+      }
+      return hashedLink;
+    }
+
+    // Handle usage-based expiration
+    if (hashedLink.maxUsageCount && hashedLink.maxUsageCount > 0) {
+      if (hashedLink.usageCount >= hashedLink.maxUsageCount) {
+        throw new Error(ErrorCode.PrivateLinkExpired);
+      }
+
+      // Increment usage count with optimistic locking
+      try {
+        await this.hashedLinksRepository.incrementUsage(hashedLink.id, hashedLink.maxUsageCount);
+      } catch (updateError) {
+        throw new Error(ErrorCode.PrivateLinkExpired);
+      }
+    }
+
+    return hashedLink;
+  }
+
+  /**
+   * Checks if a user has permission to access a link
+   * @param link - Link object with event type information
+   * @param userId - The user ID to check permissions for
+   * @returns true if user has permission, false otherwise
+   */
+  async checkUserPermissionForLink(
```

Comment:
Can we maybe reuse the logic for apps/web/lib/d/[link]/[slug]/getServerSideProps.tsx